File: frameworks/python/fastapi/Implementation/routes/template.py
from pathlib import Path

from fastapi import APIRouter, Request
from fastapi.responses import HTMLResponse
# rb:wiring template.*
from fastapi.templating import Jinja2Templates

from payloads import Payload, Payloads

# rb:wiring template.*
# FastAPI has no view layer of its own, and Jinja2Templates is what its documentation reaches for.
TEMPLATES = Jinja2Templates(directory=Path(__file__).parent.parent / "templates")


def router(p: Payloads) -> APIRouter:
    """template: the payload rendered by a Jinja2 template, which the environment compiles once."""
    routes = APIRouter()

    @routes.get("/template/small", response_class=HTMLResponse)
    async def small(request: Request) -> HTMLResponse:
        return page(request, p.small)

    @routes.get("/template/medium", response_class=HTMLResponse)
    async def medium(request: Request) -> HTMLResponse:
        return page(request, p.medium)

    return routes


def page(request: Request, payload: Payload) -> HTMLResponse:
    return TEMPLATES.TemplateResponse(request, "items.html", {"page": payload})

File: frameworks/rust/salvo/Implementation/routes/template.rs
use askama::Template;
use salvo::prelude::*;

use crate::Payloads;
use crate::payloads::Payload;

// rb:wiring template.*
/// Implementation/templates/items.html, which askama compiles into the binary when it is built.
#[derive(Template)]
#[template(path = "items.html")]
struct ItemsPage<'a> {
    body: &'a Payload,
}
// rb:end

// rb:handler template.small,template.medium
/// The page a payload renders, written as text/html.
#[derive(Clone, Copy)]
struct Page(&'static Payload);

#[handler]
impl Page {
    async fn handle(&self) -> Result<Text<String>, StatusError> {
        ItemsPage { body: self.0 }.render().map(Text::Html).map_err(|_| StatusError::internal_server_error())
    }
}
// rb:end

/// template: Salvo has no view layer, and askama is the engine Salvo's own template example
/// renders with.
pub fn router(p: &'static Payloads) -> Router {
    Router::with_path("/template")
        .push(Router::with_path("small").get(Page(&p.small)))
        .push(Router::with_path("medium").get(Page(&p.medium)))
}
